compiler/rustc_builtin_macros/src/concat_bytes.rs

```diff
@@ -0,0 +1,210 @@
+                                format!("b{}", snippet),
+                                Applicability::MachineApplicable,
+                            )
+                            .emit();
+                    } else {
+                        cx.span_err(e.span, "cannot concatenate string literals")
+                    }
+                    has_errors = true;
+                }
+                ast::LitKind::Float(_, _) => {
+                    cx.span_err(e.span, "cannot concatenate floating point literals");
+                    has_errors = true;
+                }
+                ast::LitKind::Char(_) => {
+                    if let Ok(snippet) = cx.sess.source_map().span_to_snippet(e.span) {
+                        cx.struct_span_err(e.span, "cannot concatenate character literals")
+                            .span_suggestion(
+                                e.span,
+                                "try using a byte character",
+                                format!("b{}", snippet),
+                                Applicability::MachineApplicable,
+                            )
+                            .emit();
+                    } else {
+                        cx.span_err(e.span, "cannot concatenate character literals")
+                    }
+                    has_errors = true;
+                }
+                ast::LitKind::Int(
+                    val,
+                    ast::LitIntType::Unsuffixed | ast::LitIntType::Unsigned(ast::UintTy::U8),
+                ) if val <= 255 => {
+                    if let Ok(snippet) = cx.sess.source_map().span_to_snippet(e.span) {
+                        cx.struct_span_err(e.span, "cannot concatenate numeric literals")
+                            .span_suggestion(
+                                e.span,
+                                "try wrapping the number in an array",
+                                format!("[{}]", snippet),
+                                Applicability::MachineApplicable,
+                            )
+                            .emit();
+                    } else {
+                        cx.span_err(e.span, "cannot concatenate numeric literals")
+                    }
+                    has_errors = true;
+                }
+                ast::LitKind::Int(_, _) => {
+                    cx.span_err(e.span, "cannot concatenate numeric literals");
+                    has_errors = true;
+                }
+                ast::LitKind::Bool(val) => {
+                    cx.struct_span_err(e.span, "cannot concatenate boolean literals")
+                        .span_suggestion(
+                            e.span,
+                            "try using the numeric value",
+                            (val as u8).to_string(),
+                            Applicability::MachineApplicable,
+                        )
+                        .emit();
+                    has_errors = true;
```

Comment:
Ideally we'd find a way to deduplicate the invalid element types that are invalid everywhere (both inside and outside arrays). Maybe we can have a function that emits errors if needed for an element, calling itself on any nested arrays it finds?